File: nox/src/base/token.zig
const lib = @import("../lib.zig");
const Span = lib.Span;

pub const Token = struct {
    kind: TokenKind,
    span: Span,
    line: u32,

    pub fn len(self: Self) usize {
        return self.span.len();
    }

    const Self = @This();
};

pub const TokenKind = enum {
    As,
    Break,
    Continue,
    Float,
    Function,
    Int,
    Let,
    Macro,
    Mod,
    Mut,
    New,
    Ok,
    Return,
    Struct,
    Type,
    Use,

    Exclamation,
    Semicolon,
    Ampersand,
    BackSlash,
    Question,
    ColonColon,
    Colon,
    Comma,
    Hash,
    DotDotEq,
    DotDot,
    Dot,

    LtEq,
    GtEq,
    Lt,
    Gt,
    NoEq,
    EqEq,
    Eq,
    At,

    Plus,
    Minus,
    Star,
    Slash,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBrack,
    RBrack,

    IdVal,
    IntVal,
    FloatVal,

    Document,
    Comment,
    White,
    None,
    Eof,
};

File: nox/lib/token.ml
type uToken = { kind : uTokenKind; charIdx : int; lineIdx : int; colIdx : int }
[@@deriving show { with_path = false }]

and uTokenKind =
  (* Reserved *)
  | AS
  | AND
  | BREAK
  | CONTINUE
  | ELSE
  | FALSE
  | FLOAT
  | FUNCTION
  | IF
  | INT
  | LET
  | MATCH
  | MOD
  | MUT
  | NOT
  | OR
  | PUBLIC
  | RETURN
  | UNDEFINED
  | USE
  (* Mathematical *)
  | PLUS
  | MINUS
  | STAR
  | SLASH
  (* Conditionals *)
  | EQEQ
  | EQ
  | GE
  | LE
  | GT
  | LT
  (* Symbols *)
  | SEMICOLON
  | QUESTION
  | AMPERSAND
  | COLON
  | COMMA
  | DOTDOT
  | DOT
  | BAR
  | LPAREN
  | RPAREN
  | LBRACE
  | RBRACE
  | LBRACK
  | RBRACK
  (* Values *)
  | NAMEVAL of string
  | INTVAL
  (* Other *)
  | WHITESPACE
  | DOCUMENT
  | COMMENT
  | EOF
